src/test/java/com/artipie/pypi/PypiITCase.java

```diff
@@ -150,20 +148,130 @@ void pypiPublish() throws Exception {
+            this.exec(
+                "python3", "-m", "twine", "upload", "--verbose", "--repository-url", this.url,
+                "-u", ArtipieServer.BOB.name(), "-p", ArtipieServer.BOB.password(),
+                "pypi-repo/example-pckg/*"
+            ),
+            new IsNot<>(
+                new StringContainsInOrder(
+                    new ListOf<String>(
+                        "Uploading artipietestpkg-0.0.3-py2-none-any.whl", "100%",
+                        "Uploading artipietestpkg-0.0.3.tar.gz", "100%"
+                    )
+                )
+            )
+        );
+        MatcherAssert.assertThat(
+            "Packages should not be saved in storage",
+            this.inStorage("artipietestpkg-0.0.3-py2-none-any.whl")
+                || this.inStorage("artipietestpkg-0.0.3.tar.gz"),
+            new IsEqual<>(false)
+        );
+    }
+
     @AfterEach
-    void release() {
+    void tearDown() {
         this.server.stop();
         this.cntn.stop();
     }
 
+    private void init(final boolean anonymous) throws IOException {
+        this.storage = new FileStorage(this.tmp);
+        this.server = new ArtipieServer(
+            this.tmp, "my-pypi", this.config(anonymous)
+        );
+        this.port = this.server.start();
+        this.url = String.format("http://%s:%d/my-pypi/", PypiITCase.HOST, this.port);
+        Testcontainers.exposeHostPorts(this.port);
+        this.cntn = new GenericContainer<>("python:3")
+            .withCommand("tail", "-f", "/dev/null")
+            .withWorkingDirectory("/home/")
+            .withFileSystemBind(this.tmp.toString(), "/home");
+        this.cntn.start();
+    }
+
     private String exec(final String... command) throws Exception {
         Logger.debug(this, "Command:\n%s", String.join(" ", command));
         return this.cntn.execInContainer(command).getStdout();
     }
 
+    private String pswd(final boolean anonymous) {
+        String pass = "pass";
+        if (!anonymous) {
+            pass = ArtipieServer.ALICE.password();
+        }
+        return pass;
+    }
+
+    private String urlCntn(final Optional<ArtipieServer.User> user) {
+        final String urlcntn;
+        if (user.isEmpty()) {
+            urlcntn = this.url;
+        } else {
+            urlcntn = String.format(
+                "http://%s:%s@%s:%d/my-pypi/",
+                user.get().name(), user.get().password(), PypiITCase.HOST, this.port
+            );
+        }
+        return urlcntn;
+    }
+
+    private RepoConfigYaml config(final boolean anonymous) {
+        final RepoConfigYaml yaml = new RepoConfigYaml("pypi")
+            .withFileStorage(this.tmp.resolve("repos"));
+        if (!anonymous) {
+            yaml.withPermissions(
+                new RepoPerms(
+                    new RepoPermissions.PermissionItem(
+                        ArtipieServer.ALICE.name(), new ListOf<>("*")
+                    )
+                )
```

Comment:
@genryxy `RepoPerms` now has ctor from two strings (name and password), so this expression can be simplified